crates/sema/src/typeck/mod.rs

```diff
@@ -137,3 +138,68 @@ fn same_external_params<'gcx>(gcx: Gcx<'gcx>, a: Ty<'gcx>, b: Ty<'gcx>) -> bool
     let key = |ty: Ty<'gcx>| ty.as_externally_callable_function(gcx).parameters().unwrap();
     key(a) == key(b)
 }
+
+fn check_receive_function(gcx: Gcx<'_>, contract_id: hir::ContractId) {
+    let contract = gcx.hir.contract(contract_id);
+    
+    // Libraries cannot have receive functions
+    if contract.kind.is_library() {
+        if let Some(receive) = contract.receive {
+            gcx.dcx()
+                .err("libraries cannot have receive ether functions")
+                .span(gcx.item_span(receive))
+                .emit();
+        }
+        return;
+    }
+
+    // Check for multiple receive functions
+    let mut receive_count = 0;
+    for item_id in gcx.hir.contract_item_ids(contract_id) {
+        if let Item::Function(f) = gcx.hir.item(item_id) {
```

Comment:
multiple `receive` are already rejected during lowering so you can just check contract.receive